### ISSUE TITLE
On headless Ubuntu server starting Napari crashes Python Session

### ISSUE THREAD

**Issue by roflmaostc** [napari/napari#3686]
## 🐛 Bug

When loading Napari on a headless server remotely it crashes the Python terminal.
I've seen a few issues regarding usage of it. I'm not primarily interested in using it but to prevent that the full session crashes.
Maybe I got also lost and the same issue is already existing?

## To Reproduce

```bash
fxw@FA8-24158:~$ source napari/bin/activate
(napari) fxw@FA8-24158:~$ python
Python 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import napari
>>> napari.view_image()
Validation errors in config file(s).
The following fields have been reset to the default value:

plugins -> call_order
  none is not an allowed value (type=type_error.none.not_allowed)

WARNING: could not connect to display 
WARNING: Could not load the Qt platform plugin "xcb" in "" even though it was found.
WARNING: This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, webgl, xcb.

Aborted (core dumped)
```

## Expected behavior

A error message probably? Or a warning but not Python crash.

## Environment

* Python 3.8.10
* Napari 0.4.12
* 20.04.3 LTS 

## Misc
* It's explicitly painful in Julia using [Napari.jl](https://github.com/mkitti/Napari.jl/issues/22) 😄

**Comment by Czaki** [comment]
What scenario you would like to support? 
Headless server is a server without GUI?

If the server has no GUI then what do you expect when you call the method to show the image in a window.

This message told that napari cannot find an engine to create the Window. 

Running python interpreter under xvfb may solve this problem `xvfb-run python`

**Comment by roflmaostc** [comment]
I would expect a error message but instead the whole python shell dies.
So potentially all calculation and variables are gone.

**Comment by tlambert03** [comment]
One thing you can try: 
Just uninstall qtpy in your environment, napari should still work without the gui. And you won't get qt errors like that

**Comment by roflmaostc** [comment]
That probably could work.

But the main issue is: Napari.jl (Julia wrapper) automatically installs all deps when you import it. So I didn't install anything by hand. However, during import of Napari.jl it crashes the full Julia session.

And I thought the reason is that also Napari crashes the full Python session (when opened)-
Is that fundamentally because qt kills the Python terminal? Why is it not a error but a crash?

**Comment by Czaki** [comment]
> Is that fundamentally because qt kills the Python terminal? Why is it not a error but a crash?

Yes. Qt kills process when cannot create QApplication. 

To solve your problem there is a need to add some code that will check if GUI could be created before creating a QApplication instance. But I do not do such a thing before.

**Comment by tlambert03** [comment]
> Why is it not a error but a crash?

yes, as @Czaki mentioned, this is a Qt thing.  (plenty of discussion online about this, but [this thread](https://www.riverbankcomputing.com/pipermail/pyqt/2017-July/039411.html) on the pyqt mailing list is useful).

you'll note that napari overrides sys.excepthook [here](https://github.com/napari/napari/blob/65c624cbd55102775d316568e28d284bb597f596/napari/utils/notifications.py#L227) in order to prevent Qt from aborting the process on error, but that hook doesn't get "installed" until after the QApp has already been created.  which is why it's not helping in your case.

> But the main issue is: Napari.jl (Julia wrapper) automatically installs all deps when you import it. So I didn't install anything by hand. However, during import of Napari.jl it crashes the full Julia session.

if overriding `sys.excepthook` prior to application creation doesn't work for you for any reason, then you may just need to add a step to your installation procedure (or have napari.jl do it) that allows the possibility to effectively uninstall `qtpy` after installing `napari`

**Comment by Czaki** [comment]
> you'll note that napari overrides sys.excepthook [here](https://github.com/napari/napari/blob/65c624cbd55102775d316568e28d284bb597f596/napari/utils/notifications.py#L227) in order to prevent Qt from aborting the process on error, but that hook doesn't get "installed" until after the QApp has already been created. which is why it's not helping in your case.

This prevents only from aborting on python exception. If Qt calls directly std:::exit or equivalent then I do not know a method to prevent the close of the application.

**Comment by Czaki** [comment]
Ok. I Check Qt codebase and halt is done here https://github.com/qt/qtbase/blob/3ba1a196aaac8f480c0fcb4fc005c1f4e7d7f88b/src/gui/kernel/qguiapplication.cpp#L1242

But when checking Qt documentation it looks like it is possible to prevent this failure: https://doc.qt.io/qt-5/qtglobal.html#qFatal And here is a function to register handler https://doc.qt.io/qt-5/qtglobal.html#qInstallMessageHandler, but other sources suggest that it cannot be ignored so simply.

I do test with install handler so I could manage with information, but application is still halted.

**Comment by tlambert03** [comment]
sounds like it wont help... but just fyi, superqt has a [convenience](https://github.com/napari/superqt/blob/0681f7138af27858aa6d6d5ae933989abfe371d8/src/superqt/utils/_message_handler.py#L18) around installing a message handler:

```py
from superqt import QMessageHandler

with QMessageHandler() as handler: 
    ...
```

but I think I also recall that qFatal goes below it

**Comment by willingc** [comment]
@Czaki Is this sufficiently addressed with documentation and today's CI?

**Comment by Czaki** [comment]
I will try to review of documentation in following days.
